scripts/completion/dvc.bash

```diff
@@ -58,23 +58,118 @@ _dvc_version=''
 #       x="hello"
 #       ${!x} ->  ${hello} ->  "world"
 #
-_dvc () {
-  local word="${COMP_WORDS[COMP_CWORD]}"
+_dvc ()
+{
+    local cur prev opts
+            COMPREPLY=()
+            cur="${COMP_WORDS[COMP_CWORD]}"
+            prev="${COMP_WORDS[COMP_CWORD-1]}"
+            commands="add cache checkout commit config destroy diff fetch get-url get gc
+                 import-url import init install lock metrics move pipeline pull push
+                 remote remove repro root run status unlock unprotect version"
+            global_opts="-h --help -q --quiet -v --verbose"
+            opts="-h --help -V --version"
+                 case "${COMP_CWORD}" in
+                      1)
+                         COMPREPLY=($(compgen -W "${commands} ${opts}" -- ${cur}))
+                         ;;
+                      2)
+                        case "${prev}" in
+                            add)
+                               COMPREPLY=($(compgen -W "-R --recursive -f --file --no-commit ${global_opts} $(ls ./)" -- ${cur}))
```

Comment:
Do we need to copy these over or could we use `_dvc_add` from above?

Reply:
We can use ``_dvc_add`` and use it as global variable, but I thought if anyone wants to make changes inside the function ``_dvc()`` then it will affect the global variables. Shall I change them?

Reply:
Sorry, maybe I'm not quite following. Why would anyone want to change _dvc() but not change _dvc_* ?

Reply:
Later if someone makes any change to the global variables like ``_dvc_add`` or likewise, it won't affect the autocompletion. I mean we still have issues for ``get-url`` and ``import-url``. So, if it is resolved later it won't affect the function ``_dvc()``.

Reply:
But it should update autocompletion, shouldn't it? Maybe I'm missing something, but I still don't quite understand why you copy all of this stuff again. I don't even see where old `_dvc_add` is used. It used to be used in `    local options_list="_dvc_${COMP_WORDS[1]}"	` line, but now you are just copying the contents of it here. It seems like we should either remove one or the other. Your current approach is nice because it is explicit :slightly_smiling_face:

> The previous code only showed the file names, it didn't autocomplete the commands. I have changed the code.

So your PR is meant to support subcommands? Like `dvc metrics [tab]` -> `dvc metrics show`?

>  How can I test this if it works or not?

How did you test your last patch? :slightly_smiling_face:

Reply:
All the dvc commands such as are of heirarchy 1 and the subcommands such as force, all-tags, show come under heirarchy 2. 

It is not showing any error by using python -m tests or python3 -m tests or by running prettier.

Reply:
All the dvc commands such as are of heirarchy 1 and the subcommands such as force, all-tags, show come under heirarchy 2. 

It is not showing any error by using python -m tests or python3 -m tests or by running prettier.``

Reply:
> All the dvc commands such as are of heirarchy 1 and the subcommands such as force, all-tags, show come under heirarchy 2.

But `force`, `all-tags` etc are not subcommands, they are flags. Subcommands are `show, add, remove` for `dvc metrics`, `show, list` for `dvc pipeline` and so on. And for those, subcommands are coming as level 3, not 2. 2 is indeed `--force`/other flags + `metrics`, `pipeline`,etc(so all the commands that have subcommands). E.g.
```
dvc metrics show --all-tags
[1]   [2]         [3]         [4]
```

so it should be something like
```

...
_dvc_metrics='add show remove'
_dvc_metrics_add='-t --type -x --xpath $(compgen -G *)'
_dvc_metrics_show='-t --type -x --xpath -a --all-branches -T --all-tags -R --recursive $(compgen -G *)'
_dvc_metrics_remove='$(compgen -G *)'
_dvc_pipeline='show list'
_dvc_pipeline_show=...
_dvc_pipeline_list=...

_dvc () {
  local word="${COMP_WORDS[COMP_CWORD]}"

  COMPREPLY=()

  if [ "${COMP_CWORD}" -eq 1 ]; then
    case "$word" in
      -*) COMPREPLY=($(compgen -W "$_dvc_options" -- "$word")) ;;
      *)  COMPREPLY=($(compgen -W "$_dvc_commands" -- "$word")) ;;
    esac
  elif [ "${COMP_CWORD}" -eq 2 ]; then
    local options_list="_dvc_${COMP_WORDS[1]}"

    COMPREPLY=($(compgen -W "$_dvc_global_options ${!options_list}" -- "$word"))
  elif [ "${COMP_CWORD}" -eq 2 ]; then
    local options_list="_dvc_${COMP_WORDS[1]}_${COMP_WORDS[2]}"

    COMPREPLY=($(compgen -W "$_dvc_global_options ${!options_list}" -- "$word"))
  fi

  return 0
}
```

> It is not showing any error by using python -m tests or python3 -m tests or by running prettier.

But our tests don't test bash completion. In order to test it, you need to install it as noted in https://dvc.org/doc/user-guide/autocomplete and try it out to see if it works as expected.

Reply:
As to `import-url` and `get-url`, you will need to run a convertion on `${COMP_WORDS[1]}` first that would replace all `-` with `_`, so that option_list becomes _dvc_import_url and not _dvc_import-url

Reply:
Thanks a lot. I am new to bash. But I will read these and improve.